src/internal/github/client.go

```diff
@@ -2,17 +2,77 @@ package github
 
 import (
 	"context"
-	"github.com/shurcooL/githubv4"
-	"golang.org/x/oauth2"
+	"fmt"
+	"log/slog"
 	"net/http"
+	"os"
+	"time"
+
+	"github.com/shurcooL/githubv4"
+	"golang.org/x/time/rate"
 )
 
-func NewGitHubClient(ctx context.Context, token string) *githubv4.Client {
-	return githubv4.NewClient(getGithubOauth2Client(ctx, token))
+func EnvAuthToken() (string, error) {
+	token := os.Getenv("GH_TOKEN")
+	if token == "" {
+		return "", fmt.Errorf("Expected $GH_TOKEN to be set, unable to authenticate with GitHub")
+	}
+	return token, nil
+
+}
+
+type Client struct {
+	ctx        context.Context
+	log        *slog.Logger
+	httpClient *http.Client
+	ghClient   *githubv4.Client
+}
+
+func NewClient(ctx context.Context, log *slog.Logger, token string) Client {
+	httpClient := &http.Client{Transport: &transport{
+		token:       token,
+		ctx:         ctx,
+		ratelimiter: rate.NewLimiter(rate.Every(time.Second/25), 1),
+	}}
+
+	return Client{
+		ctx:        ctx,
+		log:        log.WithGroup("github"),
+		httpClient: httpClient,
+		ghClient:   githubv4.NewClient(httpClient),
+	}
+	/* TODO
+	retryClient := retryablehttp.NewClient()
+	retryClient.RetryMax = 10
+	retryClient.Logger = nil
+	retryClient.HTTPClient = getGithubOauth2Client(token)
+	*/
+	// TODO rate limiting
+}
+
+func (c Client) WithLogger(log *slog.Logger) Client {
+	return Client{
+		ctx:        c.ctx,
+		log:        log.WithGroup("github"),
+		httpClient: c.httpClient,
+		ghClient:   c.ghClient,
+	}
 }
 
-func getGithubOauth2Client(ctx context.Context, token string) *http.Client {
-	return oauth2.NewClient(ctx, oauth2.StaticTokenSource(
-		&oauth2.Token{AccessToken: token},
-	))
+type transport struct {
+	token       string
+	ctx         context.Context
+	ratelimiter *rate.Limiter
+}
+
+func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	err := t.ratelimiter.Wait(t.ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	req = req.WithContext(t.ctx)
+	req.Header.Set("User-Agent", "OpenTofu Registry/1.0")
+	req.Header.Set("Authorization", "Bearer "+t.token)
```

Comment:
**(suggestion)** Since we don't validate that token is a non-empty string:

```suggestion
        if t.token != "" {
            req.Header.Set("Authorization", "Bearer "+t.token)
        }
```

Reply:
It is an expectation of this API that token is not an empty string (checked for in the one place that it's currently read).  With an application this small, I think this is an unnecessary check.